gtsam/slam/EssentialMatrixFactor.h

```diff
@@ -284,18 +294,21 @@ class EssentialMatrixFactor3 : public EssentialMatrixFactor2 {
    */
   Vector evaluateError(
       const EssentialMatrix& E, const double& d,
-      boost::optional<Matrix&> DE = boost::none,
-      boost::optional<Matrix&> Dd = boost::none) const override {
+      OptionalMatrixType DE, OptionalMatrixType Dd) const override {
     if (!DE) {
       // Convert E from body to camera frame
       EssentialMatrix cameraE = cRb_ * E;
       // Evaluate error
-      return Base::evaluateError(cameraE, d, boost::none, Dd);
+      return Base::evaluateError(cameraE, d, OptionalNone, Dd);
     } else {
       // Version with derivatives
       Matrix D_e_cameraE, D_cameraE_E;  // 2*5, 5*5
       EssentialMatrix cameraE = E.rotate(cRb_, D_cameraE_E);
-      Vector e = Base::evaluateError(cameraE, d, D_e_cameraE, Dd);
+	  // Had to do this since the only overloaded function EssentialMatrixFactor2
+	  // uses the type OptionalMatrixType. Which would be a pointer when we are
+	  // not using boost. There is no way to redirect that call to the top (NoiseModelFactorN)
+	  // dereference it and bring it back to the Base (EssentialMatrixFactor2)
+      Vector e = Base::evaluateError(cameraE, d, &D_e_cameraE, Dd);
```

Comment:
I'm not sure I understand this comment - I think there might be a typo ("since the only overloaded function ... uses the type") and some missing comma's, and I'm not clear what "this" and "that" is referring to.  Sorry, not trying to be aggressive, just trying to give constructive feedback on how I interpreted when reading. :)

I am indeed curious because I would have expected the entire if-else block to be replaced by something like

```c++
Vector evaluateError( /* args */) {
  Matrix D_e_cameraE, D_cameraE_E; // 2*5, 5*5

  EssentialMatrix cameraE = E.rotate(cRb_, DE ? &D_cameraE_E : 0);
  Vector e = Base::evaluateError(cameraE, d, DE ? &D_e_cameraE : 0, Dd);
  if (DE) *DE = D_e_cameraE * D_cameraE_E;
  return e;
}
```

Reply:
I have made the comment more clearer. Also by rephrasing it like this would mean cameraE will be computed even if DE is `nullptr`